backend/checkmach.ml

```diff
@@ -2254,18 +2246,93 @@ end = struct
       in
       if !changed then loop env' else env'
     in
-    if !found_unresolved
-    then (
-      let env = loop init_env in
-      Env.iter env ~f:(fun func_info v -> Func_info.update func_info v);
-      Env.print ~msg:"after fixpoint" ppf env;
-      report_unit_info ppf unit_info ~msg:"after fixpoint")
+    let env = loop init_env in
+    Env.iter env ~f:(fun func_info v ->
+        assert (Value.is_resolved v);
+        Func_info.update func_info v);
+    Env.print ~msg:"after fixpoint" ppf env
 
-  let record_unit unit_info ppf =
+  let record_unit unit_info unresolved_deps ppf =
     Profile.record_call ~accumulate:true ("record_unit " ^ analysis_name)
       (fun () ->
-        fixpoint ppf unit_info;
-        record_unit ppf unit_info)
+        report_unit_info ppf unit_info ~msg:"before fixpoint";
+        let found_unresolved = ref false in
+        let init_env =
+          Unit_info.fold unit_info ~init:Env.empty ~f:(fun func_info env ->
+              let v =
+                if Value.is_resolved func_info.value
+                then func_info.value
+                else (
+                  found_unresolved := true;
+                  Env.init_val)
+              in
+              Env.add func_info v env)
+        in
+        if !found_unresolved
+        then (
+          fixpoint ppf init_env;
+          report_unit_info ppf unit_info ~msg:"after fixpoint")
+        else assert (Unresolved_dependencies.is_empty unresolved_deps);
+        check_and_save_unit_info ppf unit_info)
+
+  (* [fixpoint_self_rec] try to resolves the common case of self-recursive
+     functions with no other unresolved dependencies, instead of waiting until
+     the end of the compilation unit to compute its fixpoint. Return the
+     unresolved callees of [func_info]. *)
+  let fixpoint_self_rec (func_info : Func_info.t) ppf =
+    let unresolved_callees = Value.get_unresolved_names func_info.value in
+    if String.Set.is_empty unresolved_callees
+       || not
+            (String.Set.equal unresolved_callees
+               (String.Set.singleton func_info.name))
+    then unresolved_callees
+    else
+      let init_env = Env.singleton func_info Env.init_val in
+      fixpoint ppf init_env;
+      report_func_info ~msg:"after self-rec fixpoint" ppf func_info;
+      String.Set.empty
+
+  (* [propagate] applies resolved values to transitive dependencies, but does
+     not resolve mutually recursive loops. *)
+  let rec propagate ~callee unit_info unresolved_deps ppf =
+    let callee_info = Unit_info.find_exn unit_info callee in
+    if Value.is_resolved callee_info.value
+       && Unresolved_dependencies.contains ~callee unresolved_deps
+    then (
+      let callers =
+        Unresolved_dependencies.get_callers ~callee unresolved_deps
+      in
+      assert (not (String.Set.is_empty callers));
```

Comment:
That was my worry to and that's why the assert is here, this condition is not trivial. I think it holds because we keep `unresolved_deps` in sync with `unit_info` in line 2321.

Reply:
(Discussed off-line.)